### ISSUE TITLE
[$500] Description field doesn't autofocus correctly in submit expense flow

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#52414]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.60-1
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @shawnborton 
**Slack conversation** (hyperlinked to channel name): [ts_external_expensify_bugs](https://expensify-ts.slack.com/archives/C049HHMV9SM/p1731409994189869)

## Action Performed:
1. Create an expense
2. Choose a recipient
3. Tap description field

## Expected Result:
We should auto-focus any text field that is the only text field on the page.

## Actual Result:
The description field is the only text field on the page, but the field doesn't quite auto-focus. The border turns green as if it was auto-focused, but you have to tap again to get the cursor to focus in the field.

## Workaround:
Unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [x] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details>

https://github.com/user-attachments/assets/c9dde090-d9ba-4785-963d-cb68fd79071f

https://github.com/user-attachments/assets/c5ea7e3f-2efe-4cf6-90c0-1b328eae51d0

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021856445363759517675</li>
        <li>Upwork Job ID: 1856445363759517675</li>
        <li>Last Price Increase: 2025-01-20</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @Kureev</details>

**Comment by Anaslancer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Description field doesn't autofocus correctly in submit expense flow

### What is the root cause of that problem?
We are trying to set a focus in [this code](https://github.com/Expensify/App/blob/3023d6016ef59d3fc8f0190f48597c41e95d3a55/src/pages/iou/request/step/IOURequestStepDescription.tsx#L83).
And it is an animated text input box.
So we should wait and set focus until the animation is ended.

### What changes do you think we should make in order to solve the problem?
We can use the `useEffect` instead of `useFocusEffect` in this below code.
https://github.com/Expensify/App/blob/3023d6016ef59d3fc8f0190f48597c41e95d3a55/src/pages/iou/request/step/IOURequestStepDescription.tsx#L79

### What alternative solutions did you explore? (Optional)
We can increase more milliseconds in `setTimeout`.
https://github.com/Expensify/App/blob/3023d6016ef59d3fc8f0190f48597c41e95d3a55/src/pages/iou/request/step/IOURequestStepDescription.tsx#L91

### Contributor details
Your Expensify account email: anasup1995@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01aff093c9a804b145

**Comment by ikevin127** [comment]
Edited by **proposal-police**: This proposal was **edited** at 2024-11-20 00:20:18 UTC.

# Proposal

### Please re-state the problem that we are trying to solve in this issue

The field doesn't quite auto-focus. The border turns green as if it was auto-focused, but then in the end you have to tap again to get the cursor to focus into the field.

### What is the root cause of that problem?

The root cause is this `IOURequestStepDescription.tsx` [block of logic](https://github.com/Expensify/App/blob/3023d6016ef59d3fc8f0190f48597c41e95d3a55/src/pages/iou/request/step/IOURequestStepDescription.tsx#L79-L93) which does not work as intended, instead only turning the input's border green without actualy focusing the input and showing the cursor / keyboard.

⚠ This is a well known issue ([ref1](https://bugs.webkit.org/show_bug.cgi?id=195884#c4), [ref2](https://github.com/JedWatson/react-select/issues/3501)) in the Webkit space where the input element appears focused but there's no visible cursor/caret is specifically related to how iOS Chrome/Safari (WebKit) implements security and user interaction guidelines.

### What changes do you think we should make in order to solve the problem?

Add `autoFocus={Browser.isMobileWebKit()}`prop to the `TextInput` (InputWrapperWithRef) [here](https://github.com/Expensify/App/blob/3023d6016ef59d3fc8f0190f48597c41e95d3a55/src/pages/iou/request/step/IOURequestStepDescription.tsx#L171-L193). This ensures we keep the current `IOURequestStepDescription.tsx` [autofocus logic](https://github.com/Expensify/App/blob/3023d6016ef59d3fc8f0190f48597c41e95d3a55/src/pages/iou/request/step/IOURequestStepDescription.tsx#L79-L93) while doubling down on autofocus only for WebKit based browsers (more context in https://github.com/Expensify/App/issues/52414#issuecomment-2487031151).

> [!note]
> I tested the first proposal's suggested solutions and none of them fixed the issue for me.

### Result videos (before / after)

<details>
<summary>iOS: mWeb Safari</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/user-attachments/assets/c48bbc1e-a480-4f34-8ff3-e6447224c53b"> | <video src="https://github.com/user-attachments/assets/dda2a31e-b49b-4b7c-a0a6-16921e7852d2"> |

</details>

**Comment by huult** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Description field doesn't autofocus correctly in submit expense flow

### What is the root cause of that problem?

We set a 300ms timeout in this block to delay the inputRef focus until after animations or screen transitions. However, 300ms was not enough for inputRef to focus properly, so this issue occurred.
https://github.com/Expensify/App/blob/main/src/pages/iou/request/step/IOURequestStepDescription.tsx#L79-L93

### What changes do you think we should make in order to solve the problem?
To resolve this issue we should `InteractionManager.runAfterInteractions` to this utility in React Native runs code only after animations and interactions have finished, ensuring that the component is ready. Some thing like that:

https://github.com/Expensify/App/blob/main/src/pages/iou/request/step/IOURequestStepDescription.tsx#L79-L93
```diff
    useFocusEffect(
        useCallback(() => {
            focusTimeoutRef.current = InteractionManager.runAfterInteractions(() => {
                if (inputRef.current) {
                    inputRef.current.focus();
                }
            });

            return () => {
                if (!focusTimeoutRef.current) {
                    return;
                }
                clearTimeout(focusTimeoutRef.current);
            };
        }, []),
    );
```

**Comment by truph01** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The field doesn't quite auto-focus. The border turns green as if it was auto-focused, but then in the end you have to tap again to get the cursor to focus into the field.

### What is the root cause of that problem?
- We already have a logic to call focus input manually in:

https://github.com/Expensify/App/blob/7256ad6eef046be3345e0b260252f6ec47fc4203/src/pages/iou/request/step/IOURequestStepDescription.tsx#L79-L93

It works well in all platform except Safari because of the following:
> We (Apple) like the current behavior and do not want programmatic focus to bring up the keyboard when you do not have a hardware keyboard attached

[Source](https://bugs.webkit.org/show_bug.cgi?id=195884#c4)

### What changes do you think we should make in order to solve the problem?
- We can add ```autoFocus={Browser.isSafari()}``` in:

https://github.com/Expensify/App/blob/7256ad6eef046be3345e0b260252f6ec47fc4203/src/pages/iou/request/step/IOURequestStepDescription.tsx#L171

to make sure the bug is fixed in safari and does not introduce regression in other platform.

### What alternative solutions did you explore? (Optional)

**Comment by truph01** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The field doesn't quite auto-focus. The border turns green as if it was auto-focused, but then in the end you have to tap again to get the cursor to focus into the field.

### What is the root cause of that problem?
- We already have a logic to call focus input manually in:

https://github.com/Expensify/App/blob/7256ad6eef046be3345e0b260252f6ec47fc4203/src/pages/iou/request/step/IOURequestStepDescription.tsx#L79-L93

It works well in all platform except IOS Safari because of the following:
> We (Apple) like the current behavior and do not want programmatic focus to bring up the keyboard when you do not have a hardware keyboard attached

[Source](https://github.com/JedWatson/react-select/issues/3501#issuecomment-535940935)

### What changes do you think we should make in order to solve the problem?
- We can add ```autoFocus={Browser.isSafari()}``` in:

https://github.com/Expensify/App/blob/7256ad6eef046be3345e0b260252f6ec47fc4203/src/pages/iou/request/step/IOURequestStepDescription.tsx#L171

to make sure the bug is fixed in safari and does not introduce regression in other platform.

### What alternative solutions did you explore? (Optional)

#### Optional:
- Beside the main solution, we can update:

https://github.com/Expensify/App/blob/7256ad6eef046be3345e0b260252f6ec47fc4203/src/pages/iou/request/step/IOURequestStepDescription.tsx#L179

to:

```
ref={inputCallbackRef}
```

with:

```
    const {inputCallbackRef} = useAutoFocusInput(true)
```

and remove [this useEffect](https://github.com/Expensify/App/blob/7256ad6eef046be3345e0b260252f6ec47fc4203/src/pages/iou/request/step/IOURequestStepDescription.tsx#L79-L93)

**Comment by sakluger** [comment]
Hey @getusha, looks like we have several proposals. Could you please take a look?

**Comment by sakluger** [comment]
Im' going to assign a new C+ to make sure we keep the urgency on this issue.

**Comment by sobitneupane** [comment]
Thanks for the proposal everyone.

I believe this issue isn’t limited to the description field. We’ve encountered similar issues in the app before, and I recall we planned to batch them into a single issue. However, it seems they’re still unresolved.

**Comment by sobitneupane** [comment]
I found the [original issue](https://github.com/Expensify/App/issues/10414), but it seems we closed it without resolving. @parasharrajat, since you were reviewing the issue, did we close it in favor of another one? Could you please share some details?

**Comment by ikevin127** [comment]
To reinforce my [proposal](https://github.com/Expensify/App/issues/52414#issuecomment-2472026500), contrary to the subsequent proposals suggesting that `autoFocus` by itself would not be enough, here's the proof that it does (on real device):

| iOS: mWeb Safari | iOS: mWeb Chrome | Android: mWeb Chrome |
| ------------- | ------------- | ------------- |
| <video src="https://github.com/user-attachments/assets/7885b525-2225-4356-847e-efbd8a8911b5"> | <video src="https://github.com/user-attachments/assets/05126060-3d52-425c-8a9a-d4267a0cf115"> | <video src="https://github.com/user-attachments/assets/75266c92-3813-4d3d-9168-cd80276c6089"> |

> [!note]
> 1. The issue is WebKit browser engine specific.
> 2. The solution I proposed only targets and is scoped specifically for this issue.
> 3. Platforms like **Web**, **Desktop** and **Native** are not below because the issue is not present there.

There's 1 downside to the solution I suggested: sometimes (not consistently, see videos above) after successfully focusing the input (caret and keyboard showing), it loses focus shortly after. Not sure why this happens but might have something to do with WebKit based on this [comment](https://bugs.webkit.org/show_bug.cgi?id=195884#c4) where they say that somewhere in the WebKit source they do this on purpose - which I suspect creates a race condition between our `autoFocus` and their `unfocus` functionality.

⬆️ Even with this downside (which doesn't happen consistently, see videos above), I think we can still consider the issue fixed and if calling it "fixed" is a stretch in this case, it's still a major improvement IMO compared to not focusing / showing the keyboard at all, 100% consistently.

##

### About the other proposals

1. There were some attempts at copying the `autoFocus` solution which I suggested first and adding `shouldDelayFocus` in hopes to make a significant enough difference to be selected but I tested that and when we add `shouldDelayFocus` it won't focus anymore which basically means it wouldn't fix our issue.
2. Another proposal suggested `autoFocus={Browser.isSafari()}` which again wouldn't fix the issue on iOS: mWeb Chrome / Firefox which also use WebKit as browser engine (mandated by Apple) since the check excludes any other WebKit based browsers other than Safari.

## Updated [proposal](https://github.com/Expensify/App/issues/52414#issuecomment-2472026500)
- added alternative solution regarding targeting `autoFocus` specifically for if `Browser.isMobileWebKit()` is `true` (all webkit based browsers)

**Comment by sakluger** [comment]
@sobitneupane how do you think we should proceed?

**Comment by sobitneupane** [comment]
@sakluger, since this issue is not limited to the Description field and occurs with every field expected to be auto-focused, I believe it would be better to address it holistically rather than handling each field individually.

**Comment by sobitneupane** [comment]
@sakluger, should we update the description to address the issue throughout the app? I’m not sure what the next step is.

**Comment by shawnborton** [comment]
I agree that we should fix this issue in all places where it might exist, so I think it makes sense to update the issue slightly.

**Comment by sakluger** [comment]
I agree we should handle everywhere. @sobitneupane or @shawnborton - would we expect all free-text fields to be auto-focused, or are there any other types of fields that also should be auto-focused?

**Comment by sobitneupane** [comment]
I believe we only autofocus the text field if it is the only text field on the page. The issue occurs with every text field that is expected to be auto-focused.

**Comment by sakluger** [comment]
Cool, that helps! I updated the expected and actual results to clarify that any text field that is the only text field on the page should auto-focus.

@Anaslancer @ikevin127 @huult @truph01 feel free to update your proposals based on the new requirements.

**Comment by huult** [comment]
## [Proposal updated](https://github.com/Expensify/App/issues/52414#issuecomment-2472263623)

- Added alternative solution 2 for the new requirement.

**Comment by shawnborton** [comment]
> would we expect all free-text fields to be auto-focused, or are there any other types of fields that also should be auto-focused?

My thinking is that we fix it for where ever we are currently using auto-focus. I don't know that we need to change the behavior of auto-focus though, in that I don't think we need to add it or remove it anywhere - let's just fix what currently exists in product.

**Comment by ikevin127** [comment]
> We should auto-focus any text field that is the only text field on the page.

I think this **Expected result** is too general of a requirement for any contributor to be able to come up with an acceptable proposal. All we can currently do as contributors is to say `We should auto-focus any text field that is the only text field on the page` in the **solution** section, which I'm sure it wouldn't be taken in consideration as a serious proposal.

I think we need **clear requirements** as the current **Expected result** involves changing / adjusting the `autoFocus` logic in multiple components which would also need a **bounty increase** given the extended scope of the issue.

##

> since this issue is not limited to the Description field and occurs with every field expected to be auto-focused, I believe it would be better to address it holistically rather than handling each field individually.

I think @sobitneupane should look up all instances where this issue occurs and let us know, since they were the one to propose the idea of extending the scope of the issue to a holistic approach. Once we have all the fields / component names we would be able to come up with a complete solution.

Otherwise, I think we should keep the scope contained to the Description field as the issue's title suggests and have the reviewer **thoroughly** test all the proposed solutions, including observations like https://github.com/Expensify/App/issues/52414#issuecomment-2487031151 and move forward with the best proposal. Then have a **separate issue** for centralizing the solution applied in this issue to cover the behaviour holistically, of course with all the fields / components where we want this to happen.

**Comment by sobitneupane** [comment]
@ikevin127 I suggested adopting a holistic approach to ensure the problem is resolved at the parent component level, if feasible. Otherwise, the same approach can be consistently applied across all relevant components.

Here are a few instances where this applies (though there are many more):
- Profile > Contact Method
- Profile >  Status
- Profile > Pronouns
- Profile > Timezone
- Details > Notes

**Comment by sakluger** [comment]
@sobitneupane do you know if there is a single auto-focus component that is broken? If so, then this seems like a reasonable ask. If there is no single auto-focus component, and fixing this would involve checking every page in the app for free-text fields, then I agree that the scope is much larger than we originally thought.

**Comment by sobitneupane** [comment]
Safari (iOS) intentionally blocks the keyboard from automatically opening in response to the `autofocus`. This behavior is widely reproducible in Safari and isn't limited to our app—it can also be observed in platforms like Facebook and Twitter.

Given this limitation, I believe we are, at best, looking for a workaround.

At this point, I cannot say for sure if the issue can be addressed by making changes to a single component or if it would require modifying every text field in the app. We previously attempted to resolve this issue in [this PR](https://github.com/Expensify/App/pull/23922), but it was ultimately closed.

**Comment by sakluger** [comment]
Ideally, we can address this by making changes to a single component. If it turns out that we need to modify every text field separately, then I would be happy to raise the bounty to $500 for the added effort (both in identifying all necessary fields as well as writing additional code).

**Comment by sakluger** [comment]
We're still waiting for an updated proposal that handles this holistically.

**Comment by sakluger** [comment]
We're still waiting for an updated proposal that handles this holistically.

**Comment by sobitneupane** [comment]
Waiting for an updated proposal.

**Comment by sakluger** [comment]
Still waiting for updated proposals.

**Comment by sakluger** [comment]
No new proposals. If we still have no proposals next week, I'll ask our agencies if they'd like to help.

**Comment by sakluger** [comment]
Holding until Monday.

**Comment by sobitneupane** [comment]
Still waiting for updated proposals.

**Comment by sakluger** [comment]
I've raised the bounty to $500. Hopefully we'll get a few new or updated proposals!

**Comment by hannojg** [comment]
I think on web it makes sense to handle the autofocus ourselves, as the `autofocus` prop on input or text input only works on page load, but not afterwards.

So I actually agree with [this proposal](https://github.com/Expensify/App/issues/52414#issuecomment-2487031151). @ikevin127 mentioned that there is a bug to why the focus isn't working reliably. While I believe this could be true, I would give it a try and see if there isn't anything else going on that might prevents it from working (so I am trying to say, maybe we can make it work 100% of the times, maybe its a bug - I'd have to double check myself).

The issue ticket https://github.com/Expensify/App/issues/54759 is primarily for mobile (however it would be nice if wee remove `shouldDelayFocus` on mobile to also unifiable remove this on web as well.).
So yes, as part of https://github.com/Expensify/App/issues/54759 I could look into this issue as well!

**Comment by sakluger** [comment]
@hannojg I believe the issue was that @ikevin127's proposal only addressed the description field - we want to solve this issue for any field that uses auto-focus.

Where we got stuck was in figuring out if there was a single component that would fix all auto-focus fields, or if we would need to identify each field and fix it separately. Do you know if there is a single component that would address all auto-focus fields?

**Comment by sakluger** [comment]
@hannojg checking in again - Do you know if there is a single component that would address all auto-focus fields?

**Comment by hannojg** [comment]
Here is what I think:

There should only be one `autoFocus` prop for our TextInput implementation, and it should _just work ™_ (talking about our [`BaseTextInput`](https://github.com/Expensify/App/tree/main/src/components/TextInput/BaseTextInput) component).

Right now additionally we have:

- a `shouldDelayFocus` prop which we use in conjunction with `autoFocus` on the `BaseTextInput` to work around some bugs (e.g. the bug we are discovering here, we would just need to use this prop to workaround those bugs)
- additionally logic in some screens / components to basically work around the same bugs that `shouldDelayFocus` aims to workaround

So my proposal is, to remove all those workarounds in different places, remove the `shouldDelayFocus` prop, and simply make sure we fix all underlying bugs that prevent `autoFocus` from working 100%.

Thats basically my proposal here:

- https://github.com/Expensify/App/issues/54759

> Do you know if there is a single component that would address all auto-focus fields?

So, my tl;dr: yes, it should be `BaseTextInput`s `autoFocus` prop that we should fix and make sure it's always working for all cases.

**Comment by sakluger** [comment]
Cool, thanks for the feedback! That's super helpful.

@sobitneupane @ikevin127 is that helpful for you?

**Comment by sobitneupane** [comment]
@sakluger I believe we should hold off on https://github.com/Expensify/App/issues/54759, as it will significantly affect the autofocus logic.

@hannojg This issue is reproducible only on iOS/mWeb and occurs across almost every component. The text field appears focused but lacks a caret and keyboard. It has been a long-standing issue.

https://private-user-images.githubusercontent.com/38435837/385399472-c5ea7e3f-2efe-4cf6-90c0-1b328eae51d0.mp4?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzczODI2NTMsIm5iZiI6MTczNzM4MjM1MywicGF0aCI6Ii8zODQzNTgzNy8zODUzOTk0NzItYzVlYTdlM2YtMmVmZS00Y2Y2LTkwYzAtMWIzMjhlYWU1MWQwLm1wND9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIwVDE0MTIzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiYjgxMzE3MTdlOTRjYjU1MWJiZmEyZWY3YzFjNzA4OWU4ZGZjZmE1MGI0Mzc3ZjllYWUxODAzZjk1ZDY2ZGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.jX48i5SCGLfa_QDfTbolEvvzPmmfMBCseolNk3vJUQc

**Comment by sakluger** [comment]
As recommended by @sobitneupane, I've put this on hold for https://github.com/Expensify/App/issues/54759 and set it to weekly to align with the priority of the hold issue.

**Comment by sakluger** [comment]
No new updates in the [hold issue](https://github.com/Expensify/App/issues/54759).

**Comment by sobitneupane** [comment]
ON HOLD

**Comment by sakluger** [comment]
No new updates in the https://github.com/Expensify/App/issues/54759.

**Comment by sakluger** [comment]
Still no updates in the hold issue.

**Comment by hannojg** [comment]
I overlooked a little that there are many related issues to what https://github.com/Expensify/App/issues/54759 tries to fix.
For example there is a new issue ticket here that basically has the same problem:

- https://github.com/Expensify/App/issues/54759

So I am bumping this up in priority and will get started working on that one finally!

**Comment by sakluger** [comment]
Thanks Hanno! Commenting for Melvin.

**Comment by sakluger** [comment]
Still on hold.

**Comment by sakluger** [comment]
Looks like the hold issue is wrapping up - it has been deployed to production and is just awaiting the regression period.

As a reminder, here's the comment where @sobitneupane recommended we put this issue on hold. @hannojg, do you think we can take this issue off hold and get it moving again?

**Comment by sakluger** [comment]
I will be OOO from May 5 - 16, so I have added another BZ member to watch over the issue while I'm out.

**Next step:** We just took the issue off hold and are waiting for advice from @sobitneupane or @hannojg.

**Comment by sobitneupane** [comment]
Still seeing the issue on iOS/Safari (mweb). @hannojg, just checking—are you planning to look into it or already on it?

**Comment by chrispader** [comment]
Hanno is not able to work on it right now as he's busy with other tasks. I will have someone to work on the project by start of next week! 🙌🏼

**Comment by sakluger** [comment]
Hey @chrispader, were you able to find anyone else to work on this one?

**Comment by chrispader** [comment]
Hey @sakluger, yes @boorad is going to work on this this week! 🙌🏼

**Comment by boorad** [comment]
After a bit more research and experimentation (and talking with @chrispader), I've started a [PR](https://github.com/Expensify/App/pull/62883
) for this issue.  The first commit has my suggested path forward.  It will require changing everywhere we use `InputWrapper` and `useAutoFocusInput` but we think this is the most holistic approach for this issue.

I have a lot to test to make sure this change will work everywhere as intended, and not disrupt existing functionality.
- verify that all uses of `useAutoFocusInput` hook are tied to `InputWrapper`
   - `inputRef` use vs `inputCallbackRef`
- figure out if we even need `useAutoFocusInput` anymore, or if `autoFocus` was the solution all along
- How do we handle `isMultiline` and scrolling with this new approach
- effects upon native and non-Mobile Webkit browsers
   - a quick `ios` test looked okay

**Comment by sakluger** [comment]
Hey @boorad, any updates on the draft PR progress?

**Comment by boorad** [comment]
@sakluger I'm back to work on it today after a bit of PTO.

Any thoughts on the suggested path forward in the PR?  It's a big patch to remove all uses of `useAutoFocusInput` hook.

TBH, I am leaning toward not touching the hook as it has a lot of code that successfully waits for animation to finish.

**Comment by ikevin127** [comment]
To follow-up on my initial [proposal](https://github.com/Expensify/App/issues/52414#issuecomment-2472026500) and @boorad's [comment](https://github.com/Expensify/App/issues/52414#issuecomment-2914114737) -> note that PR https://github.com/Expensify/App/pull/60468 was merged for issue:
- https://github.com/Expensify/App/issues/59388

where we used a new method to have focus aligned on all platforms when it comes to FAB > Start chat, but this issue is still present on **iOS: mWeb Safari** where the input gains the focus highlight style (green border) but doesn't actually display the carret / open keyboard.

I pointed out in the PR that this **iOS: mWeb Safari** focus edge case is present because as mentioned before:

> This is a well known issue ([ref1](https://bugs.webkit.org/show_bug.cgi?id=195884#c4), [ref2](https://github.com/JedWatson/react-select/issues/3501)) in the Webkit community where the input element appears focused but there's no visible cursor/caret is specifically related to how iOS Chrome/Safari (WebKit) implements security and user interaction guidelines.

##

> TBH, I am leaning toward not touching the hook as it has a lot of code that successfully waits for animation to finish.

I agree, my proposed solution still works for this Webkit specific edge case issue (see video below), not sure why this issue is still open 205 days later.

I'd say let's fix it specifically for this issue / case / component using the proposed solution and if the same issue will arise from a different component, address it in a follow-up at no additional payment for the initial PR author.

https://github.com/user-attachments/assets/f6144439-d7f6-4f87-9109-ebdda6b5fd22

cc @sobitneupane

**Comment by sobitneupane** [comment]
Thanks, @ikevin127.

I don’t think it’s enough to solve the issue only in the IOU Description. We should address it wherever it occurs, as the problem can be easily reproduced in any field that's expected to be auto-focused. Even if it requires updating every relevant text field in the app, I believe we should handle it under the same project or issue since it’s essentially the same underlying problem.

@boorad, any updates from your side? What’s your take on this?

**Comment by boorad** [comment]
@sobitneupane I have a fix that is basically @ikevin127 's proposal, but applied to all uses of `InputWrapper`.  See https://github.com/Expensify/App/pull/62883 - I'm testing all platforms to see if there are any negative effects, and going through inputs to see where `autoFocus={false}` needs to be applied.  It's definitely the minority.

**Comment by ikevin127** [comment]
As a note, I tried the `autoFocus={Browser.isMobileWebKit()}` on the inputs mentioned [here](https://github.com/Expensify/App/issues/52414#issuecomment-2947772885) (FAB -> Start chat -> Chat / Room) and looks like `autoFocus` in this case is breaking functionality previously added as part of the PR.

☝This is one of the cases in which currently adding `autoFocus` would not work as expected. I observed that this might be due to the fact that once the Start chat tab navigator is mounted (Chat / Room tabs), `autoFocus` does not work consistently, and that might be due to the fact that the tabs are already mounted after first opening and switching navigation between the 2 tabs does not necessarily constitute unmount / mount.

Noticed that `autoFocus` works best in simpler cases like this issue's description field, since that page is always mounted / unmounted, whenever its being open / closed.

##

Just a note, hope this will help with testing the PR 👍

**Comment by sakluger** [comment]
I just assigned @boorad, I realized I had never assigned them to this issue.

@boorad, from what I can see, [your draft PR](https://github.com/Expensify/App/issues/49844) appears to be stalled. How can we keep this one moving forward? Given the weekly priority label, can you please aim to provide an update at least once a week?

**Comment by boorad** [comment]
@sakluger hey, sorry about that.  The PR isn't stalled.  I ran into some time off, so hours have been scarce of late, but I worked on this last week.  I will do weekly updates based on that tag from now on.

I've been testing this wide-ranging change across a lot of screens, and have some local changes that I will push soon.  The main problem now is that in mobile WebKit, for some forms that now use this `autoFocus: true` default, React Navigation is firing a `blur` event, and the `onBlur()` handler triggers a premature form validation, causing the user to see "This field is required" and similar error messages without typing.  I'm seeing what can be done about it.  [I found this crusty issue](https://github.com/react-navigation/react-navigation/issues/11643)

An option is to not make this the default, but to go through all forms and set it to true for only certain fields.  I didn't opt for this because the thinking was most forms are a single field on a modal, but it could be a better approach.  Also, it may not fix the `onBlur()` issue, as the `autoFocus` fields may still trigger validation.

**Comment by sakluger** [comment]
Hey @boorad, any updates over the past week?

**Comment by boorad** [comment]
@sakluger hi, sorry I don't have any ideas for who could help.  @ikevin127 solved the issue originally, but the global fix is more involved.

**Comment by ikevin127** [comment]
Sorry guys, I don’t have the capacity to take this on _as expected by the reviewer_ because I don't find a global solution viable when it comes to the autofocus behaviour, as there are many cases in which a global solution would break things in many places while only fixing a few.

I would take it if I'd be allowed to go with the locally targeted approach as [previously stated](https://github.com/Expensify/App/issues/52414#issuecomment-2947772885), keeping the scope limited to this specific issue as opposed to extending the scope to the entire app, spending months trying to make it work ☠️ (can't afford that)

I know @sobitneupane [strongly disagrees](https://github.com/Expensify/App/issues/52414#issuecomment-2961826114) with that approach, but I think the complexity of their proposed approach was, in time, proven to be vastly underestimated.

Still, if they’re set on that direction and don’t want to pivot, maybe they’d be interested in taking the reins 🙌

**Comment by sobitneupane** [comment]
I could no longer reproduce the issue in the simulator, but it’s still consistently reproducible on a physical device.


>  but I think the complexity of their proposed approach was, in time, proven to be vastly underestimated.

I agree with @ikevin127 — this turned out to be more complex than it initially seemed.

>  Otherwise, the same approach can be consistently applied across all relevant components.

As suggested earlier [here](https://github.com/Expensify/App/issues/52414#issuecomment-2527710354), could we consider extending your solution to other components that are expected to auto-focus? A few of them are listed in that same comment. Let’s try to cover as many components as we reasonably can. If we run into complications for any of them, we can handle those edge cases in a separate issue.

I know this will be a bit tedious since it involves testing a lot of components.

**Comment by sakluger** [comment]
It seems like we're stalled on this one. @boorad do you think you could run with the idea that @sobitneupane suggested in the previous comment? And if not, @sobitneupane do you have any ideas of who may be good to work on this?

**Comment by sakluger** [comment]
Hey @Kureev, any luck reproducing?

**Comment by Kureev** [comment]
Hi @sakluger,

Sorry for not posting updates for the last few days. I'm currently focused on https://github.com/Expensify/App/issues/63871. As soon as I finish, this task is next in my prio list

**Comment by Kureev** [comment]
Hi @sakluger,

Sorry for delay on this one. The cycle of merging 16kb-related PRs (#63871) ended up being longer than I initially assumed. I'm planning to return to this one ASAP. There are two PRs left to finish and I'm back to this one.
